src/Razor/src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/DefaultRazorProjectHost.cs

```diff
@@ -53,6 +60,50 @@ protected override async Task OnProjectChangedAsync()
             });
         }
 
+        private void UpdateDocuments(HostProject hostProject, IEnumerable<IMSBuildItemEvaluated> projectItems)
+        {
+            var projectDirectory = Path.GetDirectoryName(hostProject.FilePath);
+
+            var currentComponentFiles = projectItems
+                .Where(item => item.Name == "Content" && item.Include.EndsWith(".razor", StringComparison.OrdinalIgnoreCase))
+                .Select(item => GetAbsolutePath(projectDirectory, item.Include))
+                .ToList();
+
+            var oldFiles = this._componentFiles;
+            var newFiles = currentComponentFiles.ToImmutableHashSet();
+
+            var addedFiles = newFiles.Except(oldFiles);
+            var removedFiles = oldFiles.Except(newFiles);
+
+            _componentFiles = currentComponentFiles;
+
+            MonoDevelop.Core.Runtime.RunInMainThread(() =>
+            {
+                foreach (var document in removedFiles)
+                {
+                    RemoveDocument(hostProject, document);
+                }
+
+                foreach (var document in addedFiles)
+                {
+                    AddDocument(hostProject, document, document.Substring(projectDirectory.Length + 1));
+                }
+            });
+        }
+
+        private string GetAbsolutePath(string projectDirectory, string relativePath)
+        {
+            if (!Path.IsPathRooted(relativePath))
+            {
+                relativePath = Path.Combine(projectDirectory, relativePath);
+            }
+
+            // normalize the path separator characters in case they're mixed
```

Comment:
Curious, given that this info comes from MSBuild when would they be mixed?

Reply:
When the absolute path is `/Users/kirill/Razor31/` and the relative path is `Pages\\Counter.razor`. MSBuild still gives us `\` everywhere.